### DISCUSSION TITLE
Management of subsets of enumeration types

### DISCUSSION THREAD

**Discussion by mpradovelasco** [OpenModelica/OpenModelica#10940] [Q&A] [answered]
Hello, 

I have a scenario where the following enumeration types may be required:

```
type chemicals = enumeration(insulin, albumin, glucose, metabolite) "All molecules";
type mcrChemicals = enumeration(insulin, albumin ) "molecules subset";
```
mcrChemicals is a enumeration mapped to the chemicals enumeration type, that keeps a subset of its values in an ordered manner.

If the following property array is defined:
`Real m[chemicals] "Property of chemicals";` , 
then the following equation could compute (for instance) the mass of small molecules:

`sum(m[ch] for ch in chemicals) - sum(m[chemicals(i)] for i in 1:Integer(chemicals.albumin))`

But if a different subtype of chemicals enumeration is required, for instance to monitorize a subset of chemicals that maps in a non-ordered way the chemicals enumeration, a mapping function is needed. For example, for the following enumeration
`type subchemicals = enumeration(insuline, glucose);`, 

the mapping function to implement could be defined by the asseveration: f(ch in subchemicals) = ch in chemicals. In this case, the previous equation could then be written as

`sum(m[ch] for ch in chemicals) - sum(m[chemicals(f(ch))] for ch in subchemicals);`

However, I wonder if Modelica allows a subtype of the class chemicals, in such a way that the following code could be written if subchemicals is a formal subtype of chemicals:

`sum(m[ch] for ch in chemicals) - sum(m[ch] for ch in subchemicals);`

Thank you for your help in advance!

**Comment by perost** [comment, marked as answer]
I don't think there's any way to achieve that in Modelica other than to write an actual mapping function. A possible alternative would be to use arrays for the subsets instead of enumerations, such as:
```Modelica
parameter chemicals subchemicals[:] = {chemicals.insuline, chemicals.glucose};
```
That would make your code work, but would of course not work if you really need them to be enumerations for other reasons.

**Comment by mpradovelasco** [comment]
Thank you for your answer @perost. Yes, it is true that the array of enumerated is a solution for many situations.
